lib/wasi/src/os/command/mod.rs

```diff
@@ -0,0 +1,104 @@
+pub mod builtins;
+
+use std::{collections::HashMap, sync::Arc};
+
+use wasmer::{FunctionEnvMut, Store};
+use wasmer_vbus::{BusSpawnedProcess, SpawnOptions};
+use wasmer_wasi_types::wasi::Errno;
+
+use crate::{bin_factory::ModuleCache, syscalls::stderr_write, WasiEnv, WasiRuntimeImplementation};
+
+/// A command available to an OS environment.
+pub trait VirtualCommand
+where
+    Self: std::fmt::Debug,
+{
+    /// Returns the canonical name of the command.
+    fn name(&self) -> &str;
+
+    /// Retrieve the command as as a [`std::any::Any`] reference.
+    fn as_any(&self) -> &dyn std::any::Any;
+
+    /// Executes the command.
+    fn exec<'a>(
+        &self,
+        parent_ctx: &FunctionEnvMut<'a, WasiEnv>,
+        path: &str,
+        store: &mut Option<Store>,
+        config: &mut Option<SpawnOptions<WasiEnv>>,
+    ) -> wasmer_vbus::Result<BusSpawnedProcess>;
+}
+
+#[derive(Debug, Clone)]
+pub struct Commands {
+    commands: HashMap<String, Arc<dyn VirtualCommand + Send + Sync + 'static>>,
+}
+
+impl Commands {
+    fn new() -> Self {
+        Self {
+            commands: HashMap::new(),
+        }
+    }
+
+    // TODO: this method should be somewhere on the runtime, not here.
+    pub fn new_with_builtins(
+        runtime: Arc<dyn WasiRuntimeImplementation + Send + Sync + 'static>,
+        compiled_modules: Arc<ModuleCache>,
+    ) -> Self {
+        let mut cmd = Self::new();
+        let cmd_wasmer =
+            builtins::cmd_wasmer::CmdWasmer::new(runtime.clone(), compiled_modules.clone());
+        cmd.register_command(cmd_wasmer);
+
+        cmd
+    }
```

Comment:
Can someone explain this TODO?

Reply:
This is due to a rather cumbersome situation where the `CmdWasmer` requires some additional context.

This would/will be fixed as part of the work in #3606